File: cpp/67. Add Binary.cpp
/*Given two binary strings, return their sum (also a binary string).

For example,
a = "11"
b = "1"
Return "100".*/

/*用了两个指针i,j分别指向a和b的末尾，然后每次取出一个字符，转为数字，若无法取出字符则按0处理，然后定义进位carry，初始化为0，
将三者加起来，对2取余即为当前位的数字，对2取商即为当前进位的值，记得最后还要判断下carry，如果为1的话，要在结果最前面加上一个1*/

class Solution {
public:
    string addBinary(string a, string b) {
        string ans = "";
        int i = a.size()-1, j = b.size()-1, carry=0;
        while (m>=0 || n>=0){
            int x = m>=0? a[i--] - '0' : 0;
            int y = n>=0? b[j--] - '0' : 0;
            int sum = x + y + carry;
            ans = to_string(sum%2) + ans;
            carry = sum / 2;
        }
        return carry==1? "1" + ans : ans;
    }
};

File: python/67. Add Binary.py
"""
67. Add Binary
Easy

Given two binary strings a and b, return their sum as a binary string.

Example 1:
Input: a = "11", b = "1"
Output: "100"

Example 2:
Input: a = "1010", b = "1011"
Output: "10101"
"""


class Solution:
    def addBinary(self, a: str, b: str) -> str:
        return '{0:b}'.format(int(a, 2) + int(b, 2))


class Solution:
    def addBinary(self, a: str, b: str) -> str:

        x, y = list(a), list(b)
        carry = 0
        res = ''

        while x or y or carry:
            if x:
                carry += int(x.pop())
            if y:
                carry += int(y.pop())

            res += str(carry % 2)
            carry //= 2

        return res[::-1]
